### DISCUSSION TITLE
uct_md.c:203  UCX  ERROR

### DISCUSSION THREAD

**Discussion by biglinn** [SSCHAcode/python-sscha#338] [Q&A] [answered]
Dear SSCHA developers,

I'm a new SSCHA user. I'm doing a Au example on SSCHA website "[Tutorial 01 - First Simulations (https://sscha.eu/Tutorials/tutorial_01_first_simulations/)".
However, I have some erros seemingly with respect to MPI. They are following:

[1717677221.254829] [n27:162451:0]         uct_md.c:203  UCX  ERROR Transport 'lx5' does not exist
[1717677221.255325] [n27:162481:0]         uct_md.c:203  UCX  ERROR Transport 'lx5' does not exist
[1717677221.255580] [n27:162451:0]         uct_md.c:203  UCX  ERROR Transport '' does not exist
[1717677221.255757] [n27:162481:0]         uct_md.c:203  UCX  ERROR Transport '' does not exist
[1717677221.256260] [n27:162481:0]         uct_md.c:203  UCX  ERROR Transport 'x5_0:1' does not exist
[1717677221.256569] [n27:162451:0]         uct_md.c:203  UCX  ERROR Transport 'x5_0:1' does not exist
[1717677221.256643] [n27:162481:0]         uct_md.c:203  UCX  ERROR Transport ':1' does not exist
[1717677221.256994] [n27:162481:0]         uct_md.c:203  UCX  ERROR Transport '' does not exist
[1717677221.257219] [n27:162451:0]         uct_md.c:203  UCX  ERROR Transport ':1' does not exist

I wonder why these erros appear.

Best regards,
Lin

**Comment by mesonepigreco** [comment]
Hi Lin,
Could you let me know how you installed the SSCHA code? Are you using Anaconda or the standard Python? Are you running on your PC or in a cluster? What is the operating system? 

This is the first time I have seen these errors, and we need more details to understand what went wrong.

**Reply by biglinn** [reply]
Thanks for your reply.
I completely follow "Easy installation through Anaconda" to install SSCHA in a cluster of linux system.

**Reply by biglinn** [reply]
I find that there is a remind after installing with Anaconda:

On Linux, Open MPI is built with UCX support but it is disabled by default.
To enable it, first install UCX (conda install -c conda-forge ucx).
Afterwards, set the environment variables
OMPI_MCA_pml=ucx OMPI_MCA_osc=ucx
before launching your MPI processes.
Equivalently, you can set the MCA parameters in the command line:
mpiexec --mca pml ucx --mca osc ucx ...

On Linux, Open MPI is built with CUDA awareness but it is disabled by default.
To enable it, please set the environment variable
OMPI_MCA_opal_cuda_support=true
before launching your MPI processes.
Equivalently, you can set the MCA parameter in the command line:
mpiexec --mca opal_cuda_support 1 ...
Note that you might also need to set UCX_MEMTYPE_CACHE=n for CUDA awareness via
UCX. Please consult UCX documentation for further details.

I dont whether erros come from this problem.

**Comment by biglinn** [comment]
Mr.
It's really the problem "Open MPI is built with CUDA awareness but it is disabled by default". When I follow its remind to install UCX and set environment variables, erros disappear.
But I meet other erros, as follows:

> Traceback (most recent call last):
>   File "/public/home/biglinn/au-sscha-test/au.py", line 2, in <module>
>     import sscha, sscha.Ensemble, sscha.SchaMinimizer, sscha.Relax, sscha.Utilities
>   File "/public/home/biglinn/miniconda3/envs/sscha/lib/python3.10/site-packages/sscha/Ensemble.py", line 29, in <module>
>     import cellconstructor.Structure
>   File "/public/home/biglinn/miniconda3/envs/sscha/lib/python3.10/site-packages/cellconstructor/Structure.py", line 27, in <module>
>     import cellconstructor.Methods as Methods
>   File "/public/home/biglinn/miniconda3/envs/sscha/lib/python3.10/site-packages/cellconstructor/Methods.py", line 22, in <module>
>     import cellconstructor.ForceTensor
>   File "/public/home/biglinn/miniconda3/envs/sscha/lib/python3.10/site-packages/cellconstructor/ForceTensor.py", line 5, in <module>
>     import cellconstructor.Phonons as Phonons
>   File "/public/home/biglinn/miniconda3/envs/sscha/lib/python3.10/site-packages/cellconstructor/Phonons.py", line 26, in <module>
>     from cellconstructor.Moro_object import Moro
>   File "/public/home/biglinn/miniconda3/envs/sscha/lib/python3.10/site-packages/cellconstructor/Moro_object.py", line 29, in <module>
>     import matplotlib.pyplot as plt
>   File "/public/home/biglinn/miniconda3/envs/sscha/lib/python3.10/site-packages/matplotlib/pyplot.py", line 55, in <module>
>     import matplotlib.colorbar
>   File "/public/home/biglinn/miniconda3/envs/sscha/lib/python3.10/site-packages/matplotlib/colorbar.py", line 19, in <module>
>     from matplotlib import _api, cbook, collections, cm, colors, contour, ticker
>   File "/public/home/biglinn/miniconda3/envs/sscha/lib/python3.10/site-packages/matplotlib/contour.py", line 15, in <module>
>     from matplotlib.backend_bases import MouseButton
>   File "/public/home/biglinn/miniconda3/envs/sscha/lib/python3.10/site-packages/matplotlib/backend_bases.py", line 49, in <module>
>     from matplotlib import (
>   File "/public/home/biglinn/miniconda3/envs/sscha/lib/python3.10/site-packages/matplotlib/text.py", line 16, in <module>
>     from .font_manager import FontProperties
>   File "/public/home/biglinn/miniconda3/envs/sscha/lib/python3.10/site-packages/matplotlib/font_manager.py", line 1585, in <module>
>     fontManager = _load_fontmanager()
>   File "/public/home/biglinn/miniconda3/envs/sscha/lib/python3.10/site-packages/matplotlib/font_manager.py", line 1580, in _load_fontmanager
>     json_dump(fm, fm_path)
>   File "/public/home/biglinn/miniconda3/envs/sscha/lib/python3.10/site-packages/matplotlib/font_manager.py", line 965, in json_dump
>     with cbook._lock_path(filename), open(filename, 'w') as fh:
>   File "/public/home/biglinn/miniconda3/envs/sscha/lib/python3.10/contextlib.py", line 135, in __enter__
>     return next(self.gen)
>   File "/public/home/biglinn/miniconda3/envs/sscha/lib/python3.10/site-packages/matplotlib/cbook.py", line 1878, in _lock_path
>     raise TimeoutError("""\
> TimeoutError: Lock error: Matplotlib failed to acquire the following lock file:
>     /public/home/biglinn/.cache/matplotlib/fontlist-v390.json.matplotlib-lock
> This maybe due to another process holding this lock file.  If you are sure no
> other Matplotlib process is running, remove this file and try again

I find that I perform the commond "pip install cellconstructor **python-sscha** tdscha" when I install SSCHA.
However, the script import "sscha" instead of "python-sscha". So I guess whether this leads to erros.

**Reply by mesonepigreco** [reply, marked as answer]
Are you submitting the job in parallel? 

The error here is from matplotlib, as unveiled from the last lines of your error:
```
TimeoutError: Lock error: Matplotlib failed to acquire the following lock file:
/public/home/biglinn/.cache/matplotlib/fontlist-v390.json.matplotlib-lock
This maybe due to another process holding this lock file. If you are sure no
other Matplotlib process is running, remove this file and try again
```

You can manually remove the file, it says. Indeed, if you run the job in parallel, it is possible that multiple process created that file, thus leading to the error. This should not occur, but it is most likely due to an error of how MPI is configured.

Try removing that file (if any) and rerun it with 1 processor.

**Reply by biglinn** [reply]
Yes, I run in parallel. However, there is not a file named "fontlist-v390.json.matplotlib-lock" but a file named "fontlist-v390.json" in the folder ".cache/matplotlib/".
I take a try to rerun it with 1 processor.

**Reply by mesonepigreco** [reply]
Ok, let me know if on a single processor it works. 
It is possible that the parallelization is not correctly set, and all the jobs execute exactly the same things, conflicting and giving the previous error. It should not happen, but this depends on the openmpi configuration.

**Reply by biglinn** [reply]
It's really that erros result from not correctly setting parallelization as you say.
I remove file in the folder ".cache/matplotlib/" what ever its name is "fontlist-v390.json.matplotlib-lock" or not. Then I preform the job on a single proceesor, and it works. Moreover, I perform the job on multiple processors and it also works.

**Comment by biglinn** [comment]
@mesonepigreco 
I find a question. In the tutorial, sscha use a dynamic matrix named "harmonic_dyn", "sscha_T300_dyn", etc. These files don't have extensions. And according to their contents, it seems they are text file.
But I preformed QE calculations and output dynamic matrix as "PREFIX.dyn.xml" by setting "fildyn".
So I wonder if I can output dynamic matrix as the same as sscha by setting "fildyn".

**Reply by biglinn** [reply]
SSCHA calculate anharmonic dynamic matrices based on harmonic ones. 
When SSCHA do minimization with a supercell, therefore, does it need harmonic matrices of corresponding supercell, or just unitcell?

**Reply by diegomartinez2** [reply]
You can use the dynamic matrices of the unit cell and set the supercell on SSCHA `sscha.Ensemble.Ensemble(dyn,T,SUPERCELL)`.

**Reply by biglinn** [reply]
So does it means that I can set supercell of any size to calculate their anharmonicity as long as I have harmonic dynamic matrices of the unit cell?

**Reply by diegomartinez2** [reply]
The harmonic dynamic matrices are used as a starting point for the SSCHA anharmonic calculations. The main difference will be the number of steps the SSCHA will need to converge the anharmonic results.

**Reply by biglinn** [reply]
I see. Thanks!
Moreover, I want to perform SSCHA+DFT. According to the tutorial, I need to excute the command "mpirun -np 50 script.py > output".
But in the script, there is a line "command = 'mpirun -np 8 pw.x -npool 1 -i PREFIX.pwi > PREFIX.pwo'  ".
Will these two mpi numbers conflict? Should I keep these two numbers the same?

**Reply by diegomartinez2** [reply]
They are different things; the first one runs the script.py and the second runs the Quantum espresso. Usually the code runs several QE calculations in a separate cluster. In that case the script is run locally with 50 processes in parallel, and then it launches the QE calculations (in a cluster) with another number of processes (here 8). 
Note: For the QE calculations use a '-npool' that is a divisor of '.np' for example with -np=8 -npool can be 1,2,4 or 8.

**Reply by biglinn** [reply]
Does it mean that the script is split into 50 processes, then each one of the 50 launches the QE calculations with 8 processes.
If so, since DFT have most of computational cost, should I set the number of script processes to a small value and set the number of QE processes to a large one?

**Reply by diegomartinez2** [reply]
No, the script uses the 50 processes on the SSCHA calculations, the QE calculations are done between the SSCHA steps, so it uses QE for the calculation of the forces, etc. and then it uses the 50 processes for the free energy minimization until it finds that it cannot keep using this forces and launches QE again to get the new forces, etc. until the minimization condition is meet. That the self-consistent part of the "stochastic self-consistent harmonic approximation".

**Reply by biglinn** [reply]
I see. In a separate iteration, SSCHA minimization is done after QE calculations, and then a second iteration is performed.

Supposing that I use SLURM system to apply for 50 processors, and the script does SSCHA minimization with these 50. Between SSCHA minimization steps, QE calculations use 8 of these 50.
Is that the case?

**Reply by diegomartinez2** [reply]
That's the case only if you do not use the cluster function for calling the QE from SSCHA. If you make a manually QE call from SSCHA then you will use the processors declared in the SLURM run. On the other hand if you use the cluster functionality, then the QE calculations are processed as a serie of different SLURM calls and are independent from the main SSCHA script.

Check the [http://sscha.eu/Tutorials/tutorial_02_advanced_submission/](http://sscha.eu/Tutorials/tutorial_02_advanced_submission/)

**Reply by biglinn** [reply]
That's it!
 What I want is that the main SSCHA script can perform several SLURM tasks for QE.
Thanks a lot for your help!  I'll study this tutorial well.

**Reply by mesonepigreco** [reply]
@biglinn 
Usually, the sscha works on a local computer that can connect to a cluster via an SSH connection, and then automatically submit via SLURM the espresso calculation while running all the remaining calculations locally.

The new version (1.5), which will be released very soon, introduces a new mode that enables running the SSCHA on the cluster with a long serial job. Then, that job proceeds by submitting other jobs with the espresso calculations directly to the cluster. This second mode is very useful if the SSH connection to the cluster is difficult to maintain (e.g., we do not have a workstation, or the connection to the cluster must occur via a 2FA protocol that requires human interaction).

**Reply by biglinn** [reply]
Oh, that may be my problem! I exactly submit the SSCHA script on a cluster. So I'll study the available mode.
And I'm definitely looking forward to this second mode! It's friendly for users who directly work on a cluster like me.

**Comment by biglinn** [comment]
@mesonepigreco and @diegomartinez2 
There is an issue that confuses me. After I excuted the script to generate ensemble and sample configurations as follows, all configurations are supercells(here 5*5*5). I dont konw which command created supercell in this script.

This script "ensenmble.py" is as follows:
```python
import cellconstructor as CC, cellconstructor.Phonons
import sscha, sscha.Ensemble
import numpy as np

# Fix the seed so that we all generate the same ensemble
np.random.seed(0)

# Load the dynamical matrix
dyn = CC.Phonons.Phonons("../ktao3_harmonic_dyn", nqirr=19)
dyn.Symmetrize()
#[ apply here the needed changes to dyn ]

# Prepare the ensemble
temperature = 300 # 300 K
ensemble = sscha.Ensemble.Ensemble(dyn, temperature)

# Generate the ensemble
number_of_configurations = 10
ensemble.generate(number_of_configurations)

# Save the ensemble into a directory
save_directory = "data"
population_id = 1
ensemble.save(save_directory, population_id)
```

**Reply by mesonepigreco** [reply]
Hi,
The "supercell size" is defined by the q-mesh sampling of your dynamical matrix.
In your case:
```python
dyn = CC.Phonons.Phonons("../ktao3_harmonic_dyn", nqirr=19)
```
Loads the dynamical matrix ``ktao3_harmonic_dyn`` on 19 irreducible q points. The q_points stored inside these files are commensurate with a 5x5x5 supercell (which is big for an ab-initio calculation!).

If this is one of your first calculations, I suggest starting with a more modest size (e.g. 2x2x2), get the results for the 2x2x2, and then progressively increase the size of the supercell. In this way, you can use the final result of the 2x2x2 calculation as a starting point for bigger supercell and further reduce the computational cost.

You can either start with a harmonic dynamical matrix on the 2x2x2 mesh, or interpolate the 5x5x5 you already have down to a 2x2x2 with the command:

```python
# Interpolate from the 5x5x5 to the 2x2x2 supercell
new_dyn = dyn.Interpolate( (5,5,5), (2,2,2))
```
Then use your new_dyn to generate the ensemble.

**Reply by biglinn** [reply]
Oh, I see.
So, can I get dynamical matrix with a small q-mesh first, and progressively increase the size of the supercell in different iterations?
You say that interpolating from the 5x5x5 to the 2x2x2 supercell is okay.  Does Interpolating from the small to the large supercell if I want increase the supercell size?

**Reply by mesonepigreco** [reply]
Yes.
However, there are tricks to increase the precision of the interpolation. 
Anyway, the best would be to run a 2x2x2 harmonic calculation and start from that one.
Then interpolate from the 2x2x2 -> 3x3x3, run again the SSCHA on the 3x3x3 interpolated and check if the results you got with the 2x2x2 are correct or they need to be further converged.
And proceed until convergence.

**Reply by biglinn** [reply]
About convergence, 
Generally, do we compare the change of phonons to decide whethe results converge?
Like this figure in the tutorial:
![frequency_minim](https://github.com/SSCHAcode/python-sscha/assets/155951912/4d43d6de-4201-4e4c-b22d-7305b0f235cf)

**Reply by biglinn** [reply]
Moreover, are minimization steps and iterations the same thing?

**Reply by mesonepigreco** [reply]
Yes, you should also check if the gradients reach zero within the stochastic noise.

**Reply by biglinn** [reply]
Hi, @mesonepigreco 
I have performed ensemble generation and QE calculations. And I have a problem in energy minimization.
I use the scirpt "minim.py" form tutorial, as follows:
import sscha, sscha.Ensemble, sscha.SchaMinimizer
import sscha.Utilities
import cellconstructor as CC, cellconstructor.Phonons

```
POPULATION = 1

dyn = CC.Phonons.Phonons("harmonic_dyn/ktao3_harmonic_dyn", 19)
ensemble = sscha.Ensemble.Ensemble(dyn, 0)
ensemble.load("data_and_job", population = POPULATION, N=10)

minim = sscha.SchaMinimizer.SSCHA_Minimizer(ensemble)
minim.init()

# Save the minimization details
ioinfo = sscha.Utilities.IOInfo()
ioinfo.SetupSaving("minim_{}".format(POPULATION))


minim.run(custom_function_post = ioinfo.CFP_SaveAll)
minim.finalize()
minim.dyn.save_qe("dyn_mat/sscha_dyn_population{}_".format(POPULATION))
```
However, the program reports a erro:

> Traceback (most recent call last):
>   File "/public/home/biglinn/sscha-dft-ktao3/minim.py", line 9, in <module>
>     ensemble.load("data_and_job", population = POPULATION, N=10)
>   File "/public/home/biglinn/miniconda3/envs/sscha/lib/python3.10/site-packages/sscha/Ensemble.py", line 627, in load
>     structure.coords += disp
> ValueError: operands could not be broadcast together with shapes (625,3) (40,3) (625,3)
Why does this happen?

**Reply by biglinn** [reply]
`ensemble = sscha.Ensemble.Ensemble(dyn, 0)`
Ensembles generation has been finished. So what is the function of this line?

**Reply by mesonepigreco** [reply]
There is a mismatch between the size of the data loaded in the ensemble (the supercell size) and the one of the dynamical matrix used to initialize the ensemble.
Some of the files have a different number of atoms (40 instead of 625). Are the ensembles generated for different dynamical matrices? 

The line you mention initializes the ensemble with the provided dynamical matrix and at the given temperature (0 K). This also sets the number of atoms expected when you call the load function and could be the reason behind the mismatch. Try to use the same dynamical matrix employed to generate the ensemble.

Anyway, consider opening a new discussion for follow-up questions, as they are no longer related to the title and could get confusing for other people.

**Reply by biglinn** [reply]
Okay, thanks! I'll open a new discussion if I have another problem.

**Reply by biglinn** [reply]
Hi, @mesonepigreco

As you said before, the line `ensemble = sscha.Ensemble.Ensemble(dyn, 0)` aims to initializes the ensemble with dyn and give temperature.
So this temperature should be the same with the temperature used to generate ensembles?
Why is it set to 0K in the tutorial? As follows:

![image](https://github.com/user-attachments/assets/8077279a-0dff-4bb5-9e01-75cf12f9b648)

**Reply by biglinn** [reply]
In other words, does this temperature play a role in minimization?

**Reply by biglinn** [reply]
The ensembles have be loaded by `ensemble.load`(namely u_populationX_X.dat), I wonder if this temperature setting have some effects.

**Reply by biglinn** [reply]
Hi, @mesonepigreco

Besides, the `ensemble = sscha.Ensemble.Ensemble(dyn, 0)` also needs to be initialized when calculating 3rd-order IFCs. As follows:
```
dyn = CC.Phonons.Phonons("sscha_dyn_population2_", 4)
ensemble = sscha.Ensemble.Ensemble(dyn, 0)
ensemble.load("/public/home/biglinn/mote2/mote2-dyn/441-no_soc/data_and_job", population = 2, N=20)
ensemble.update_weights(ensemble.dyn_0, 300)

dyn_hessian, d3_tensor = ensemble.get_free_energy_hessian(include_v4 = False, get_full_hessian = True, return_d3 = True)
np.save("d3.npy", d3_tensor)
```
In this situation, does the temperature in `ensemble = sscha.Ensemble.Ensemble(dyn, 0)` play a role?

**Reply by mesonepigreco** [reply]
Yes, as it is used to compute the displacement displacement correlation function

**Reply by biglinn** [reply]
Thanks! I thought that it's just used to initialize `sscha.Ensemble.Ensemble` to use some function of this class.
